### PR TITLE
remove unnecessary second premise from stdlib's set_diff_nodup

### PR DESCRIPTION
**Pull request by palmskog**
The statement of `ListSet.set_diff_nodup` is currently:
```coq
Lemma set_diff_nodup l l' : NoDup l -> NoDup l' -> NoDup (set_diff l l').
```
However, the second premise (`NoDup l'`) is unnecessary, which was first pointed out to me by @traiansf. Here, I remove this premise so that we instead get the stronger lemma:
```coq
Lemma set_diff_nodup l l' : NoDup l -> NoDup (set_diff l l').
```

- [x] Added **changelog**.

### CODE REVIEW

**Comment by Alizter** [comment]
@coqbot run full ci

File: theories/Lists/ListSet.v
Comment (SkySkimmer) [inline theories/Lists/ListSet.v:457]: Aren't we supposed to do a deprecation dance for this sort of change?

**Comment by Zimmi48** [comment]
If this lemma was used somewhere outside Coq's CI, it could be difficult to write a backward-compatible fix, couldn't it? If you have a strategy in mind, could you document it in the changelog?

**Comment by olaure01** [comment]
This looks like an instance of the [strengthening problem](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/Deprecation.20deletion.20overwriting). Do we have a strategy for dealing with that.

**Comment by andres-erbsen** [comment]
The change looks desirable by itself. The CI failure appears unrelated.

Backwards compatibility is indeed a concern, but requiring a rename+deprecate-remove-rename+deprecate-remove dance across 4 Coq versions to land changes like this seems burdensome to the point of deterring them. As pointed out on Zulip, it also does not help developments that maintain compatibility with more than two Coq versions. Instead, I propose following the suggestion from @JasonGross and adding the previous version of this lemma to `theories/Compat` for 8.17, fully qualified like [here](https://github.com/coq/coq/blob/110921a449fcb830ec2a1cd07e3acc32319feae6/theories/Compat/Coq813.v#L20). This might still break developments that reference the previous version by its fully qualified name, but it would be possible to fix this unlikely case in a backwards-compatible way by using import-order shadowing prioritize the lemma from `Compat` over the one from `ListSet` if both exist.

@coq/stdlib-maintainers wdyt?

**Comment by Zimmi48** [comment]
That sounds like a good strategy to me. We are not making enough use of the `-compat` mechanism and I agree that the rename+deprecate-remove-rename+deprecate-remove dance is really too difficult and makes changes to the stdlib painfully slow. If that's the chosen strategy, it might make sense to document it in our contributing guide (there's already a section on contributing to the stdlib that might need to be updated to reflect the current practice).

**Comment by palmskog** [comment]
I followed the suggestion from @andres-erbsen and modified `Compat/Coq817.v` to provide the old lemma statement for `set_diff_nodup`. However, I'm not sure what the deprecation message should be, and not sure what best practices are when one can't directly deprecate a lemma (I added the prefix `_old` and deprecated a notation). Maybe someone can take a look.

File: theories/Compat/Coq817.v
Comment (andres-erbsen) [inline theories/Compat/Coq817.v:21]:
```suggestion
      Local Lemma Private_set_diff_nodup :
```

File: theories/Compat/Coq817.v
Comment (andres-erbsen) [inline theories/Compat/Coq817.v:31]:
```suggestion
      Notation set_diff_nodup := Private_set_diff_nodup.
```

File: theories/Compat/Coq817.v
Comment (andres-erbsen) [inline theories/Compat/Coq817.v:28]:
```suggestion
  auto using set_diff_nodup.
```

File: theories/Compat/Coq817.v
Comment (andres-erbsen) [reply theories/Compat/Coq817.v:21]: https://github.com/coq/coq/issues/16949

**Review by andres-erbsen — COMMENTED** [review]
The deprecation message looks good to me. I added some suggestions, but my sense is that we should move ahead and merge after fixing the CI failure unless something new is brought up.

**Comment by andres-erbsen** [comment]
The CI failure says
```
File "./Coq816.v", line 13, characters 0-33:
Error: Trying to mask the absolute name "Coq.Lists.ListSet"!
[masking-absolute-name,deprecated]
```

I guess we can't have the full path in the compat file after all. How about just dropping `Coq`?

**Comment by JasonGross** [comment]
> The CI failure says
> 
> ```
> File "./Coq816.v", line 13, characters 0-33:
> Error: Trying to mask the absolute name "Coq.Lists.ListSet"!
> [masking-absolute-name,deprecated]
> ```
> 
> I guess we can't have the full path in the compat file after all. How about just dropping `Coq`?

Just do `Local Set Warnings "-masking-absolute-name".`

**Comment by SkySkimmer** [comment]
Needs  to be export set otherwise users get the error no?

**Comment by palmskog** [comment]
@SkySkimmer I now have the following in `Compat/Coq817.v`, but I still get the error Jason flagged up:
```coq
Export Set Warnings "-masking-absolute-name".
(* ... code that masks a name ... *)
Export Set Warnings "masking-absolute-name".
```
What's going wrong here? Is it possible to get this warning ignored, or should I drop the `Coq` part of the namespace?

**Comment by JasonGross** [comment]
> What's going wrong here?

The warning is re-triggered on `Import`.  I was under the impression that effects are sequenced, but it might be the case that the nametab is updated all at once, after(?) playing all the effects?  Or it might be that warnings are added after names?  In any case, making it Global should fix this, as long as Import occurs after Require

**Comment by andres-erbsen** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
>The warning is re-triggered on Import.

on Import or on Require?

**Comment by JasonGross** [comment]
> > The warning is re-triggered on Import.
> 
> on Import or on Require?

In general both/either, but in this case, it should be `Import`.  The absolute name is not masked until you say "let me access the symbol without the `Coq817` prefix"

**Review by andres-erbsen — APPROVED** [review]
I intend to merge this in a day or two unless there are objections.

The compat solution here seems decent to me. It may not be the final word in the backwards-compatibility story for standard-library lemmas, but I think it is good enough for this simple change and worth trying out in general.

File: doc/changelog/11-standard-library/16926-set_diff_nodup-premise.rst
Comment (Zimmi48) [inline doc/changelog/11-standard-library/16926-set_diff_nodup-premise.rst:3]: I recommend it documenting it the way that users are supposed to use it:

```suggestion
  with `-compat 8.17` providing the old version of :g:`set_diff_nodup` for compatibility
```

**Review by Zimmi48 — APPROVED** [review]

**Comment by andres-erbsen** [comment]
@coqbot merge now

**Pull request merged**
